File: src/youinc_ledger/ledger_pipeline/pipeline.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable

from youinc_ledger.models import JournalTransaction, Posting, RawTransaction
from youinc_ledger.persistence_layer.db import LedgerDatabase
from youinc_ledger.rules_router.rules import RouteDecision, RulesRouter


@dataclass(frozen=True)
class PipelineResult:
    seen: int = 0
    raw_inserted: int = 0
    posted: int = 0
    skipped_pending: int = 0
    skipped_duplicate: int = 0
    skipped_zero_amount: int = 0
    errors: tuple[str, ...] = ()

    def add_error(self, error: str) -> "PipelineResult":
        return PipelineResult(
            seen=self.seen,
            raw_inserted=self.raw_inserted,
            posted=self.posted,
            skipped_pending=self.skipped_pending,
            skipped_duplicate=self.skipped_duplicate,
            skipped_zero_amount=self.skipped_zero_amount,
            errors=(*self.errors, error),
        )


class LedgerPipeline:
    def __init__(
        self,
        database: LedgerDatabase,
        router: RulesRouter,
        discard_pending: bool = True,
    ) -> None:
        self.database = database
        self.router = router
        self.discard_pending = discard_pending

    def reclassify_existing_journals(self) -> PipelineResult:
        result = PipelineResult()
        for transaction in self.database.fetch_journaled_raw_transactions():
            result = self._replace(result, seen=result.seen + 1)
            try:
                if transaction.amount_cents == 0:
                    result = self._replace(
                        result, skipped_zero_amount=result.skipped_zero_amount + 1
                    )
                    continue
                journal_transaction = self._build_journal_transaction(transaction)
                self.database.replace_journal_transaction(journal_transaction)
                result = self._replace(result, posted=result.posted + 1)
            except Exception as exc:  # noqa: BLE001 - keep batch running and report per transaction.
                result = result.add_error(str(exc))
        return result

    def process_payloads(self, payloads: Iterable[dict]) -> PipelineResult:
        result = PipelineResult()
        for payload in payloads:
            result = PipelineResult(
                seen=result.seen + 1,
                raw_inserted=result.raw_inserted,
                posted=result.posted,
                skipped_pending=result.skipped_pending,
                skipped_duplicate=result.skipped_duplicate,
                skipped_zero_amount=result.skipped_zero_amount,
                errors=result.errors,
            )
            try:
                transaction = RawTransaction.from_akahu_payload(payload)
                inserted = self.database.upsert_raw_transaction(transaction)
                if inserted:
                    result = self._replace(result, raw_inserted=result.raw_inserted + 1)

                if transaction.is_pending and self.discard_pending:
                    self.database.mark_raw_skipped(transaction.idempotency_hash, "pending")
                    result = self._replace(result, skipped_pending=result.skipped_pending + 1)
                    continue

                if transaction.amount_cents == 0:
                    self.database.mark_raw_skipped(transaction.idempotency_hash, "zero_amount")
                    result = self._replace(
                        result, skipped_zero_amount=result.skipped_zero_amount + 1
                    )
                    continue

                if self.database.journal_exists(transaction.idempotency_hash):
                    result = self._replace(result, skipped_duplicate=result.skipped_duplicate + 1)
                    continue

                journal_transaction = self._build_journal_transaction(transaction)
                if self.database.insert_journal_transaction(journal_transaction):
                    result = self._replace(result, posted=result.posted + 1)
                else:
                    result = self._replace(result, skipped_duplicate=result.skipped_duplicate + 1)
            except Exception as exc:  # noqa: BLE001 - keep batch running and report per transaction.
                result = result.add_error(str(exc))
        return result

    @staticmethod
    def _replace(result: PipelineResult, **changes: int) -> PipelineResult:
        values = {
            "seen": result.seen,
            "raw_inserted": result.raw_inserted,
            "posted": result.posted,
            "skipped_pending": result.skipped_pending,
            "skipped_duplicate": result.skipped_duplicate,
            "skipped_zero_amount": result.skipped_zero_amount,
            "errors": result.errors,
        }
        values.update(changes)
        return PipelineResult(**values)

    def _resolve_route(self, transaction: RawTransaction) -> RouteDecision:
        """Manual per-transaction classifications win over rule/nzfcc routing."""
        override = self.database.get_manual_classification(transaction.idempotency_hash)
        if override is not None:
            target_account, memo = override
            return RouteDecision(
                target_account=target_account,
                rule_id="manual:override",
                memo=memo,
                matched_by="manual",
            )
        return self.router.route(transaction)

    def _build_journal_transaction(self, transaction: RawTransaction) -> JournalTransaction:
        account_mapping = self.router.account_mapping_for(transaction.account_id)
        route = self._resolve_route(transaction)
        amount_cents = abs(transaction.amount_cents)

        if transaction.amount_cents > 0:
            postings = (
                Posting(
                    account=account_mapping.ledger_account, side="debit", amount_cents=amount_cents
                ),
                Posting(account=route.target_account, side="credit", amount_cents=amount_cents),
            )
        else:
            postings = (
                Posting(account=route.target_account, side="debit", amount_cents=amount_cents),
                Posting(
                    account=account_mapping.ledger_account, side="credit", amount_cents=amount_cents
                ),
            )

        journal_transaction = JournalTransaction(
            external_id=transaction.idempotency_hash,
            transaction_date=transaction.settlement_date[:10]
            if transaction.settlement_date
            else transaction.transaction_date,
            description=transaction.description,
            source_account_id=transaction.account_id,
            status=transaction.status,
            rule_id=route.rule_id,
            postings=postings,
        )
        journal_transaction.validate_balanced()
        return journal_transaction

File: frontend/src/server/ledger-engine/pipeline.ts
import { fromAkahuPayload, type RawTransaction } from "./rawTransaction";
import { RulesRouter, type RouteDecision } from "./rulesRouter";
import {
  posting,
  validateBalanced,
  type JournalTransaction,
  type Posting,
} from "./journal";

/**
 * TypeScript port of youinc_ledger.ledger_pipeline.pipeline.LedgerPipeline
 * (process_payloads path) — P2 ledger port. Proven at parity against the
 * Python engine by tests/golden/fixtures/journal_balancing.json (see
 * pipeline.golden.test.ts).
 *
 * The pipeline is pure orchestration over a persistence port (LedgerStore).
 * The Python engine drove SQLite directly; here the store is an interface so
 * the multi-tenant target can back it with the (deferred) Supabase DAL, while
 * the golden test supplies an in-memory store that mirrors the exact SQLite
 * semantics the fixture pins (upsert insert/duplicate, journal existence,
 * manual-classification override).
 *
 * reclassify_existing_journals is intentionally not ported — no fixture pins
 * it, and it depends on a read path (fetch_journaled_raw_transactions) that
 * belongs to the DAL slice, not this one.
 */

/** Persistence port. Method contracts mirror LedgerDatabase exactly:
 *  - upsertRawTransaction returns true only when the hash was newly inserted.
 *  - insertJournalTransaction returns true when inserted, false if a journal
 *    with the same external_id already exists.
 *  - getManualClassification returns [targetAccount, memo] or null. */
export interface LedgerStore {
  upsertRawTransaction(transaction: RawTransaction): boolean;
  markRawSkipped(externalId: string, reason: string): void;
  journalExists(externalId: string): boolean;
  insertJournalTransaction(transaction: JournalTransaction): boolean;
  getManualClassification(externalId: string): [string, string | null] | null;
}

export interface PipelineResult {
  seen: number;
  rawInserted: number;
  posted: number;
  skippedPending: number;
  skippedDuplicate: number;
  skippedZeroAmount: number;
  errors: ReadonlyArray<string>;
}

function getErrorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

export class LedgerPipeline {
  constructor(
    private readonly store: LedgerStore,
    private readonly router: RulesRouter,
    private readonly discardPending: boolean = true,
  ) {}

  processPayloads(payloads: Iterable<Record<string, unknown>>): PipelineResult {
    const result = {
      seen: 0,
      rawInserted: 0,
      posted: 0,
      skippedPending: 0,
      skippedDuplicate: 0,
      skippedZeroAmount: 0,
      errors: [] as string[],
    };

    for (const payload of payloads) {
      result.seen += 1;
      try {
        const transaction = fromAkahuPayload(payload);
        const inserted = this.store.upsertRawTransaction(transaction);
        if (inserted) result.rawInserted += 1;

        if (transaction.isPending && this.discardPending) {
          this.store.markRawSkipped(transaction.idempotencyHash, "pending");
          result.skippedPending += 1;
          continue;
        }

        if (transaction.amountCents === 0) {
          this.store.markRawSkipped(transaction.idempotencyHash, "zero_amount");
          result.skippedZeroAmount += 1;
          continue;
        }

        if (this.store.journalExists(transaction.idempotencyHash)) {
          result.skippedDuplicate += 1;
          continue;
        }

        const journalTransaction = this.buildJournalTransaction(transaction);
        if (this.store.insertJournalTransaction(journalTransaction)) {
          result.posted += 1;
        } else {
          result.skippedDuplicate += 1;
        }
      } catch (error: unknown) {
        // Keep the batch running and report per transaction (parity with the
        // Python engine's broad per-payload except).
        result.errors.push(getErrorMessage(error));
      }
    }

    return result;
  }

  /** Manual per-transaction classifications win over rule/nzfcc routing. */
  private resolveRoute(transaction: RawTransaction): RouteDecision {
    const override = this.store.getManualClassification(transaction.idempotencyHash);
    if (override !== null) {
      const [targetAccount, memo] = override;
      return {
        targetAccount,
        ruleId: "manual:override",
        memo,
        matchedBy: "manual",
      };
    }
    return this.router.route(transaction);
  }

  private buildJournalTransaction(transaction: RawTransaction): JournalTransaction {
    const accountMapping = this.router.accountMappingFor(transaction.accountId);
    const route = this.resolveRoute(transaction);
    const amountCents = Math.abs(transaction.amountCents);

    // Positive amount (money in): source account DEBITed, target CREDITed.
    // Negative amount (money out): target DEBITed, source CREDITed.
    // Both legs use abs(amount_cents).
    const postings: [Posting, Posting] =
      transaction.amountCents > 0
        ? [
            posting(accountMapping.ledgerAccount, "debit", amountCents),
            posting(route.targetAccount, "credit", amountCents),
          ]
        : [
            posting(route.targetAccount, "debit", amountCents),
            posting(accountMapping.ledgerAccount, "credit", amountCents),
          ];

    const journalTransaction: JournalTransaction = {
      externalId: transaction.idempotencyHash,
      transactionDate: transaction.settlementDate
        ? transaction.settlementDate.slice(0, 10)
        : transaction.transactionDate,
      description: transaction.description,
      sourceAccountId: transaction.accountId,
      status: transaction.status,
      ruleId: route.ruleId,
      postings,
    };
    validateBalanced(journalTransaction);
    return journalTransaction;
  }
}
